pkg/virt-handler/node-labeller/kvm-caps-info-plugin.go

```diff
@@ -0,0 +1,263 @@
+	logger        *log.FilteredLogger
+}
+
+type capability struct {
+	Name      string
+	Extension uintptr
+	MSR       int
+}
+
+func NewCapScanner() *CapScan {
+	capScan := new(CapScan)
+	capScan.logger = log.DefaultLogger()
+	return capScan
+}
+
+var CapsDesc = []capability{
+	{
+		Extension: uintptr(C.CapHyperv),
+		Name:      "base",
+	},
+	{
+		Extension: uintptr(C.CapHypervTime),
+		Name:      "time",
+	},
+	{
+		Extension: uintptr(C.CapHypervVpIndex),
+		MSR:       HV_X64_MSR_VP_INDEX,
+		Name:      "vpindex",
+	},
+	{
+		Extension: uintptr(C.CapHypervTlbflush),
+		Name:      "tlbflush",
+	},
+	{
+		Extension: uintptr(C.CapHypervSendIPI),
+		Name:      "ipi",
+	},
+	{
+		Extension: uintptr(C.CapHypervSynic),
+		MSR:       HV_X64_MSR_SCONTROL,
+		Name:      "synic",
+	},
+	{
+		Extension: uintptr(C.CapHypervSynic2),
+		MSR:       HV_X64_MSR_SCONTROL,
+		Name:      "synic2",
+	},
+	{
+		MSR:  HV_X64_MSR_TSC_FREQUENCY,
+		Name: "frequencies",
+	},
+	{
+		MSR:  HV_X64_MSR_RESET,
+		Name: "reset",
+	},
+	{
+		MSR:  HV_X64_MSR_VP_RUNTIME,
+		Name: "runtime",
+	},
+	{
+		MSR:  HV_X64_MSR_STIMER0_CONFIG,
+		Name: "synictimer",
+	},
+	{
+		MSR:  HV_X64_MSR_REENLIGHTENMENT_CONTROL,
+		Name: "reenlightenment",
+	},
+}
+
+func availableMsrs(fd uintptr) ([]uint32, error) {
+
+	// Find our list of MSR indices.
+	// A page should be more than enough here,
+	// eventually if it's not we'll end up with
+	// a failed system call for some reason other
+	// than E2BIG (which just says n is wrong).
+	msrIndices := make([]byte, 4096, 4096)
+	msrs := make([]uint32, 0, 0)
+	for {
+
+		_, err := kvmIOCtl(
+			fd,
+			uintptr(C.IoctlGetMsrIndexList),
+			uintptr(unsafe.Pointer(&msrIndices[0])))
+		if err == syscall.E2BIG {
+			// The nmsrs field will now have been
+			// adjusted, and we can run it again.
+			continue
+		} else if err != 0 {
+			return nil, err
+		}
+
+		// We're good!
+		break
+	}
+
+	// Extract each msr individually.
+	for i := 0; ; i++ {
+		// Is there a valid index?
+		var index C.__u32
+		e := C.msr_list_index(
+			unsafe.Pointer(&msrIndices[0]),
+			C.int(i),
+			&index)
+
+		// Any left?
+		if e != 0 {
+			break
+		}
+
+		// Add this MSR.
+		msrs = append(msrs, uint32(index))
+	}
+	return msrs, nil
+}
+
+func hasCapExtension(fd uintptr, extension uintptr) bool {
+
+	res, err := kvmIOCtl(fd, uintptr(C.IoctlCheckExtension), extension)
+	if res != 1 || err != 0 {
+		return true
+	}
+
+	return false
+}
+
+func kvmIOCtl(fd, flag, arg uintptr) (res uintptr, err syscall.Errno) {
+	res, _, err = syscall.Syscall(syscall.SYS_IOCTL,
+		fd,
+		flag,
+		arg)
+	return
+}
+
+func (s *CapScan) populateCaps() {
+	msrs, err := s.getKVMMSRs()
+	if err != nil {
+		return
+	}
+	s.supportedMSRS = make(map[uint32]bool)
+	for _, kvmMsr := range msrs {
+		s.supportedMSRS[kvmMsr] = true
+	}
+}
+
+func (s *CapScan) getKVMMSRs() ([]uint32, error) {
+	devkvm, err := os.OpenFile("/dev/kvm", syscall.O_RDWR|syscall.O_CLOEXEC, 0)
+	if err != nil {
+		devkvm.Close()
```

Comment:
It is not needed to close the file on error. When error is returned, the file is not opened.